### PR TITLE
Create bounty program

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

**Comment by sungam3r** [comment]
@Shane32 Do you come up with all this yourself or use well-known sources?

**Comment by Shane32** [comment]
The general guidelines are similar to the OBS Project bounty program.  They seemed to fit exceedingly well with the goals and process I previously outlined.  The legal terms I rewrote based on Microsoft's bounty program.  What's not here is internal rules/processes we may wish to define for us maintainers.

**Comment by sungam3r** [comment]
@Shane32 Just for information. I have never read this PR so far and waiting for it to stabilize 👍 . Apparently, the next weeks I will spend on studying the federation mechanism since I promised to deal with #3144.

**Comment by Shane32** [comment]
> waiting for it to stabilize

I’ll request a review at that time.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:27]:
```suggestion
7. Pull requests are merged into the master or develop branch.
8. Developer(s) submit invoice to the [GraphQL.NET Open Collective](https://opencollective.com/graphql-net)
   for bounty amount, confirmed by GraphQL.NET team members.
    * Bounty status changed to **Claimed**
9. Payment disbursed to developer(s) by the Open Source Collective from GraphQL.NET funds.
```

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:13]: Do you mean that all these statuses need to be formalized as labels?

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:23]:
```suggestion
7. Pull requests are merged into the `master` or `develop` branch.
```

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:12]:
```suggestion
   should be prioritized, or a bug that is in need of being researched and fixed. These publications take place in the form of GitHub issues.
```

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:18]:
```suggestion
    * Bounty status of issue changed to **In Progress**
```

and in other places

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:27]:
```suggestion
9. Payment disbursed to developer(s) by the Open Source Collective from GraphQL.NET funds.
    * Bounty status of issue changed to **Paid**
```
?

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:14]: What to do with "instant" PRs without even writing proposal about implementation or anything?

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:27]: Do we need a confirmation about payment from the PR author?

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:44]: _notify_ should be clarified

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:47]: > for the amount of the bounty

I think there should be section describing this term. Where to find that amount? Is that amount constant or may be changed over time? Or even may be withdrawn?

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:30]: I do not understand how this section differs from `# Bounty Process Flowchart`. Yes, they differ of course, but both describe about the same.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:60]: > will be released for another team to work on

So, `In Progress` -> `Accepting Proposals` transition here? If so that should be mentioned in Bounty Process Flowchart section, right?

Comment (sungam3r) [inline BOUNTY.md:60]: task -> issue ?

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:63]: I see different terms : task (issue) vs project. Do you mean GitHub's projects feature here?

Comment (sungam3r) [reply BOUNTY.md:63]: Maybe we can use it https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects as a flow for bounty issue?

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:13]: I was going to create a few badges, yes, that correspond to these statuses.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:27]: Not sure exactly what you mean.  I'm sure us maintainers will handle the payment or the PR author will inform us if we forget!  The badge/label is really so we can indicate the number of paid bounties within the readme easily.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:14]: The bounty program isn't designed for such a purpose.  We should be collaborating periodically on what we should apply a bounty to, and then create an issue for the few tasks and mark them as having a bounty.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:44]: I'm not sure how.  They just post in the PR just like all other communication.  Seems pretty obvious to me.

Want to add something more specifically like "request PR review from a maintainer"?

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:47]: The bounty amount will be described in the issue and depend on the bounty.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:60]: yes

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:63]: > I see different terms : task (issue) vs project. Do you mean GitHub's projects feature here?

No issue is what I meant.

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:27]: I mean the fact of transferring money to the author's account.

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:14]: Then it is better to explicitly say that such PRs do not fall into the program.

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:44]: Yes. Notify sounds not formal.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:47]:
```suggestion
9. [Submit an expense through Open Collective](https://opencollective.com/graphql-net/expenses/new) for the amount of the bounty. The bounty amount will be described in the issue and depend on the bounty.
```

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:60]: 👍  Let's use consistent naming across document.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:63]:
```suggestion
If you no longer wish to work on an issue that has been granted to you, please notify the GraphQL.NET team to reset the
```

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:64]:
```suggestion
bounty status as **Accepting Proposals**.
```

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:27]: OpenCollective is responsible for the actual payment.  joemcbride just needs to approve the invoice.  There's nothing additional that we need to write here.  I'm sure if the author has an issue they will bring it up.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:44]:
```suggestion
6. Once code is complete, remove draft status and request a review from the GraphQL.NET maintainers.
```

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:32]: They are not much different.  Just written from different points of view.  Seemed fine to me, since we don't have much else to say.

**Comment by Shane32** [comment]
@sungam3r I did another pass at the entire document taking into consideration the items you had commented on.  The status are now simplified to 3 states corresponding to these GitHub labels on issues:
1. **`bounty`** - Just an RFP
2. **`bounty`** + **`in progress`** - Bounty has been granted to a dev.
3. **`bounty-paid`** - Bounty has been paid

Using these labels allows us to use badges within the readme file to:
1. Show outstanding bounties, which include those that have been granted but not finished
2. Show completed bounties

Also I attempted to unify on 'project', although often 'bounty' seemed more applicable.  Try to give it a cold read and see if it looks a bit better.

Finally, while we could merge the "flowchart" section with the "how to claim a bounty" section, it seemed better to me to leave as-is.

Please review when you have time - thanks!

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:28]: OK. I expected such an answer and only wanted to clarify.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:28]:
```suggestion
9. Payment disbursed to developer(s) by the Open Source Collective from GraphQL.NET funds, i.e. Open Collective is
   responsible for the actual payment.
```

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:32]: I am inclined to agree. Over time, we may find perfect formulations and descriptions.

File: README.md
Comment (sungam3r) [inline README.md:3]: dead link?

**Comment by sungam3r** [comment]
Did you forget to add `in progress` label?

**Comment by sungam3r** [comment]
I want to finish review today.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:7]:
```suggestion
Routine bug fixes and similar pull requests like fixing typos or outdated pieces of documentation do not fall into this program at this time.
```

Though I believe that fixing [confirmed bugs](https://github.com/graphql-dotnet/graphql-dotnet/issues?q=is%3Aopen+is%3Aissue+label%3Abug) should be paid as well.

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:25]:
```suggestion
7. Pull requests are merged into the `master` or `develop` branch. Also, if necessary, it can be branches for old versions.
```

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:29]:
```suggestion
    * The GitHub issue is labeled with **`bounty-paid`** and the aforementioned **`bounty`** and **`in progress`** labels are removed.
```

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:44]: >  may submit competing proposals at this time as well.
> If your proposal is accepted

So.. can other competing proposals arise after this one was marked as accepted (in progress)?

File: BOUNTY.md
Comment (sungam3r) [inline BOUNTY.md:96]: What if someone wants to post a PR closing RFP and at the same time he is under sanctions? Dev **may** do it without claims to participate into program. Right? So what about other competing proposals/PRs? Will all proposals be considered on equal terms, regardless of whether a dev falls under conditions of the Program or not?

File: README.md
Comment (sungam3r) [inline README.md:6]: Already exists in the table below.

File: README.md
Comment (sungam3r) [inline README.md:19]:
```suggestion
| :heavy_dollar_sign: [Get paid for contributing!](BOUNTY.md) :heavy_dollar_sign: | [![GitHub issues by-label](https://img.shields.io/github/issues-raw/graphql-dotnet/graphql-dotnet/bounty?color=blue&label=open%20bounties)](https://github.com/graphql-dotnet/graphql-dotnet/issues?q=is%3Aopen+is%3Aissue+label%3Abounty) [![GitHub closed issues by-label](https://img.shields.io/github/issues-closed-raw/graphql-dotnet/graphql-dotnet/bounty-paid?color=blue&label=bounty-paid)](https://github.com/graphql-dotnet/graphql-dotnet/issues?q=is%3Aclosed+is%3Aissue+label%3Abounty-paid)
```

?

File: README.md
Comment (sungam3r) [inline README.md:351]:
```suggestion
Please help us to express gratitude to those individuals who devote time and energy to contributing to this project by supporting their efforts in a tangible means.
```

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:28]:
```suggestion
9. Payment disbursed to developer(s) by the Open Source Collective from GraphQL.NET funds;
   i.e. Open Collective is responsible for the actual payment.
```

File: README.md
Comment (sungam3r) [reply README.md:3]: I thought so.

**Comment by Shane32** [comment]
> Did you forget to add `in progress` label?

I didn't forget, but neither did I add it yet.  I'll do so,

File: README.md
Comment (Shane32) [reply README.md:19]: No; it is correct.  It shows two "shields", one displayed as (1) "open bounties" and (2) one displayed as "paid bounties" which searches for (1) open issues marked with `bounty` and (2) closed issues marked with `bounty-paid`.  Preview the markdown to see how it will look.

File: README.md
Comment (Shane32) [reply README.md:6]: Yes it does.  I'm unifying with the server project readme, showing the current version along the top line of the readme.

File: README.md
Comment (sungam3r) [reply README.md:19]: 🤦‍♂️ of course...

File: README.md
Comment (sungam3r) [reply README.md:6]: OK. Then add contributors badge before repo size.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:97]: Sure.  I mean, it's pretty explicit that the entire bounty program is at the whims of the maintainers, and we can do whatever we want.  Honestly, we don't have enough contributors to worry about it, and most of the "flowchart" is just standard practice of any outstanding issue.  I often outline a PR before posting it when I believe you may have a concern with it, or when I am looking for advice before spending time on the implementation.  Or I may post a PR, as in #3414 and then subject to discussion rewrite it like #3417 .  So here if someone says they want to work on a issue and forgo the bounty, sure we just assign it like any other.  They finish the work and we skip the payment step.  I don't think we need to write a special paragraph for this; it's common sense and we just handle the scenario if and when it arises.

Note: that's a copy and paste from Microsoft.  Probably some legal reason why it's required, just like the age thing -- and since joemcbride is located in the US and will be approving the bounties, it seemed an appropriate safeguard for his sake.  It's not like we are going to be reviewing contributor's eligibility claims.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:45]: The deadline written in the bounty program should help a lot preventing these situations.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:25]:
```suggestion
7. Pull requests are merged into the appropriate branch (typically `master` or `develop`).
```

Comment (Shane32) [reply BOUNTY.md]: Please see alternate language suggested.

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:97]: ok

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md:45]: "do what we want" maybe obvious for us but not so obvious for someone who try to contribute. Here my only intent is to make things as clear as possible. Your explanation is good and I would like to move these words into document (even in the form of non-normative notice). If you do not want, then okay.

File: BOUNTY.md
Comment (sungam3r) [reply BOUNTY.md]: 👍 good

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:7]: It's not a bad idea.  We would need to determine a structure for how this works in practice.  Having a standing price for bug fixes may be to little or too much depending on the issue.

I do like that it would encourage users of the library to fix bugs themselves rather than waiting for us to do so.

However, we would need to address how this works for us maintainers.  While I think the bounty program as written works with us maintainers (as we all agree on the bounties beforehand), bugs that you and I work on are usually spur-of-the-moment issues.  Sometimes it's just one line of code.  Sometimes it's a complex problem.  Do we just set a price and go with it per-bug?  What if the bug has multiple facets that need to be addressed?

Another option is that bug fixes are available for contributors but not maintainers.

I have not thought about it much, but I'd say in order to encourage community bug fixes, I would suggest a higher bounty price for community fixes, versus a lower price for us maintainers.  It is also easier for us to fix bugs than them, as we have greater knowledge of the codebase, so from one perspective, it's fair.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:46]: Hmm..  I don't think we should say anything here.  We can write a separate document for internal procedures that we can use as guidelines in these scenarios, if the need arises.  This separate document should only be visible to maintainers.

File: README.md
Comment (Shane32) [reply README.md:6]: It's already there, down one line, and there's no more room on the top line.

image

I grouped contributors with the number of weekly/etc commits (which we don't have in the server repo because of the few number of commits).

**Comment by Shane32** [comment]
image
image

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md]: If we do add a 'bug bounty', we should state that fixes must include a test for the bug.

Comment (sungam3r) [reply BOUNTY.md]: > However, we would need to address how this works for us maintainers. 

For me it does not work at all :) ... _You are a resident of any countries under U.S. sanctions_ ...

> Another option is that bug fix bounties are available for contributors but not maintainers.

I'm fine with this option.

> If we do add a 'bug bounty', we should state that fixes must include a test for the bug.

Good. Agree.

File: README.md
Comment (sungam3r) [reply README.md:6]: ok

**Comment by sungam3r** [comment]
I finished here. Resolve or close pending conversations. Waiting for review from @joemcbride .

**Review by sungam3r — APPROVED** [review]

File: BOUNTY.md
Comment (Shane32) [inline BOUNTY.md:86]: @joemcbride Along with the rest of this document, please review this 'Legal' section when you have a minute.

These terms were paraphrased from MS's bounty program here:
https://www.microsoft.com/en-us/msrc/bounty-terms?rtc=1

I selected restrictions from Microsoft's text based on these basis:
1. Age -- due to [COPPA](https://www.ftc.gov/legal-library/browse/rules/childrens-online-privacy-protection-rule-coppa)
2. Individual researcher, etc -- for copyright
3. Resident of countries with US sanctions -- could be a legal requirement
4. MIT license requirements -- for copyright
5. Similarity waiver -- for copyright
6. No guaranteed compensation -- so we can revoke a bounty if needed
7. Submission is own work -- for copyright

I think most of these make sense, but the one regarding US sanctions we may not need.  Specifically, even if there is a US law that applies (which there very likely may not be), technically it is Open Collective that has both received donations and is paying the bounties.  You're simply approving the payment.  (Correct??)  I'd recommend removing that restriction and let Open Collective deal with any issues that may arise.  I really don't want to exclude anyone from this program, and specifically @sungam3r currently falls into this category.

We could also argue that the age restriction falls in the same category.  But as children's privacy is a hot button in almost every country, I'd leave it in there.  I'm sure any 14yo contributors can get their parent to apply for the bounty if need be!

Note that Microsoft has additional restrictions for government employees and Microsoft employees that I didn't feel were applicable here.

Final note: the "submission is own work" restriction could be an issue if someone continues another person's work.  But at the same time, if that was the case, should they be paid a bounty?  I guess I'd just leave it in there for now, and if an issue arises, we can handle it at that time.

Thanks in advance!

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md]: #3430

File: BOUNTY.md
Comment (joemcbride) [reply BOUNTY.md:86]: > I think most of these make sense, but the one regarding US sanctions we may not need. Specifically, even if there is a US law that applies (which there very likely may not be), technically it is Open Collective that has both received donations and is paying the bounties. You're simply approving the payment. (Correct??) I'd recommend removing that restriction and let Open Collective deal with any issues that may arise. I really don't want to exclude anyone from this program, and specifically @sungam3r currently falls into this category.

I agree with this sentiment, that this should be handled at the Open Collective level. Since I'm no lawyer and I'm sure those restrictions change and there are probably a lot of nuances. It would be difficult for the maintainers to try to make that determination.

**Review by joemcbride — APPROVED** [review]
This looks good to me. One suggestion I have is to have a template for a GitHub issue that has a checkbox and a link to this Bounty readme to have the requestor confirm that they have read the document.

File: BOUNTY.md
Comment (Shane32) [inline BOUNTY.md:100]:
```suggestion
```

**Comment by Shane32** [comment]
> This looks good to me. One suggestion I have is to have a template for a GitHub issue that has a checkbox and a link to this Bounty readme to have the requestor confirm that they have read the document.

Yes; if nothing else, a template for us maintainers to use when creating bounties, so all bounty issues are well structured.

File: BOUNTY.md
Comment (Shane32) [reply BOUNTY.md:86]: Done.

**Comment by Shane32** [comment]
@sungam3r @joemcbride If you like, give it one last look; otherwise as both of you have approved it, I'll merge it tonight.

**Comment by Shane32** [comment]
> These bounty hunter bots went totally out of control...

@gao-artur I think we should revoke all bounties and remove the bounty program.  We only ever had a couple people actually work on a bounty until now, and if we want AI quality work we don't need third parties to do it.  It takes more time to properly review work done by an AI than to do it myself.  I'm swamped now so I have neither reviewed these PRs nor worked on the docs myself.

**Pull request merged**
